corecomponent/androidcomponent/src/main/res/xml/shortcuts.xml

```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="utf-8"?>
+<shortcuts
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:tools="http://schemas.android.com/tools"
+    >
+    <shortcut
+        android:shortcutId="floor_map"
+        android:icon="@mipmap/ic_launcher"
+        android:shortcutShortLabel="@string/floor_map_shortcut_short_label1"
+        tools:targetApi="25">
+        <intent
+            android:action="android.intent.action.VIEW"
+            android:data="https://droidkaigi.jp/2020/floormap" />
```

Comment:
Thanks!
The current implementation, show also chrome.
Can you add android:targetPackage?

```
      android:targetPackage="com.example.myapplication"
      android:targetClass="com.example.myapplication.ComposeActivity"
```

Reply:
@takahirom 

Thanks for your comment!
I tried adding "android: targetPackage".
```
        <intent
            android:action="android.intent.action.VIEW"
            android:targetPackage="io.github.droidkaigi.confsched2020"
            android:data="https://droidkaigi.jp/2020/floormap" />
```
But it still shows chrome. 

Also, if you add "android: targetClass", the shortcut will not work and the message "This app is not installed" will be displayed.
```
        <intent
            android:action="android.intent.action.VIEW"
            android:targetPackage="io.github.droidkaigi.confsched2020"
            android:targetClass="io.github.droidkaigi.confsched2020.MainActivity"
            android:data="https://droidkaigi.jp/2020/floormap" />
```

In this case, are the specified contents of "android: targetPackage" and "android: targetClass" wrong?

Reply:
I guess that's because the package of debug build differs from `io.github.droidkaigi.confsched2020`. You may need to use a manifest placeholder` like the following.

```
android:targetPackage="${applicationId}"
android:targetClass="${applicationId}.MainActivity"
```

Reply:
Yes. When the app is debug build, We should use `io.github.droidkaigi.confsched2020.debug`

Reply:
Sorry,
I used the manifest placeholder, but the problem has not been resolved.
```
        <intent
            android:action="android.intent.action.VIEW"
            android:targetPackage="${applicationId}"
            android:targetClass="${applicationId}.MainActivity"
            android:data="https://droidkaigi.jp/2020/floormap" />
```

Also, I thought that the manifest placeholder could not be referenced, so I specified `io.github.droidkaigi.confsched2020.debug` directly, but the problem still persisted.
```
        <intent
            android:action="android.intent.action.VIEW"
            android:targetPackage="io.github.droidkaigi.confsched2020.debug"
            android:targetClass="io.github.droidkaigi.confsched2020.debug.MainActivity"
            android:data="https://droidkaigi.jp/2020/floormap" />
```

Reply:
I'm sorry, it was my mistake.

`targetClass` should be like `android:targetClass="io.github.droidkaigi.confsched2020.MainActivity"` because it must be a fully qualified domain name, not depend on `applicationId` at runtime.

Reply:
Even though the application id kinda dynamic, FQDNs of classes won't change.

Reply:
Unfortunately, this problem has not been resolved.

However, since the error log when the application does not start was confirmed, it will be explained together with the implementation details.

from: corecomponent/androidcomponent/src/main/res/xml/shortcuts.xml
```
        <intent
            android:action="android.intent.action.VIEW"
            android:targetPackage="${applicationId}"
            android:targetClass="io.github.droidkaigi.confsched2020.MainActivity"
            android:data="https://droidkaigi.jp/2020/floormap" />
```


Error log
```
2020-01-28 23:29:16.031 1405-4604/? E/LauncherAppsService: Couldn't start activity, code=-92
2020-01-28 23:29:16.043 1941-1941/? E/BaseDraggingActivity: Unable to launch. tag=WorkspaceItemInfo(id=-1 type=DEEPSHORTCUT container=-1 screen=-1 cell(-1,-1) span(1,1) minSpan(1,1) rank=0 user=UserHandle{0} title=フロアマップ) intent=Intent { act=android.intent.action.MAIN cat=[com.android.launcher3.DEEP_SHORTCUT] flg=0x10200000 pkg=io.github.droidkaigi.confsched2020.debug cmp=io.github.droidkaigi.confsched2020.debug/io.github.droidkaigi.confsched2020.MainActivity bnds=[404,308][1009,462] (has extras) }
    android.content.ActivityNotFoundException: Shortcut could not be started
        at android.content.pm.LauncherApps.startShortcut(LauncherApps.java:1173)
        at android.content.pm.LauncherApps.startShortcut(LauncherApps.java:1138)
        at com.android.launcher3.shortcuts.DeepShortcutManager.startShortcut(SourceFile:1)
        at com.android.launcher3.BaseDraggingActivity.startShortcutIntentSafely(SourceFile:10)
        at com.android.launcher3.BaseDraggingActivity.startActivitySafely(SourceFile:9)
        at com.android.launcher3.Launcher.startActivitySafely(SourceFile:4)
        at com.android.launcher3.touch.ItemClickHandler.startAppShortcutOrInfoActivity(SourceFile:13)
        at com.android.launcher3.touch.ItemClickHandler.onClickAppShortcut(SourceFile:13)
        at com.android.launcher3.touch.ItemClickHandler.a(SourceFile:6)
        at c.a.a.p.c.onClick(Unknown Source:2)
        at android.view.View.performClick(View.java:7259)
        at com.android.launcher3.shortcuts.DeepShortcutTextView.performClick(SourceFile:9)
        at android.view.View.performClickInternal(View.java:7236)
        at android.view.View.access$3600(View.java:801)
        at android.view.View$PerformClick.run(View.java:27892)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
```

I am considering two approaches based on the above.

1. Investigate the cause of the error log.
1. Reimplement with Dynamic shortcut.

And I'll see if the "Reimplement with Dynamic Link" method works.

Reply:
> Investigate the cause of the error log.

I found a related article on the above.
However, it doesn't seem to be resolved.
https://stackoverflow.com/questions/43494350/android-7-1-app-shortcuts-couldnt-start-activity?rq=1
https://stackoverflow.com/questions/42301830/activitynotfoundexception-shortcut-could-not-be-started

> Reimplement with Dynamic shortcut.

I implemented it with the following contents.
However, it seems that the screen to select Chrome and the application is displayed.

```
        val map = ShortcutInfo.Builder(this, "map")
            .setShortLabel(getString(R.string.floor_map_shortcut_short_label1))
            .setIcon(Icon.createWithResource(this, R.mipmap.ic_launcher))
            .setIntent(Intent(Intent.ACTION_VIEW,
                Uri.parse("https://droidkaigi.jp/2020/floormap")))
            .build()
```

Reply:
How about use  Dynamic shortcut and set ComponentName explicitly?
```kotlin
val map = ShortcutInfo.Builder(this, "map")
    .setShortLabel(getString(R.string.floor_map_shortcut_short_label1))
    .setIcon(Icon.createWithResource(this, R.mipmap.ic_launcher))
    .setIntent(Intent(Intent.ACTION_VIEW,
            Uri.parse("https://droidkaigi.jp/2020/floormap")
        ).setComponent(ComponentName(this, MainActivity::class.java))
    ).build()
```

Reply:
`${applicationId}` cannot be used in shortcut.xml because `${applicationId}` are only replaced in the manifest.
I don't know how to reference applicationId from resource xml...

Reply:
Thank you for your great work! I also have investigated and published a short article. http://jmatsu.hatenablog.com/entry/2020/02/03/013802
Appreciate and I'm very sorry for confusing you.

One of the solutions is to use dynamic shortcuts for now and [a way which @ohmae suggested](https://github.com/DroidKaigi/conference-app-2020/pull/467#discussion_r373848951) sounds good.

Reply:
@syo-sa1982

Reply:
@ohmae @jmatsu 
Thanks for the advice!
I'll try it today or tomorrow. 🙏

Reply:
@ohmae @jmatsu
I was able to solve it. Thank you very much. 🙏 

@takahirom 
Sorry for being late, please review again. 🙏 

<img src="https://user-images.githubusercontent.com/1960883/73847692-230cdc00-486a-11ea-927e-8fd690e2277e.gif" width="300" />